src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,113 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
+@SuppressWarnings("unchecked")
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int capacity;
+    private T[] arrayList;
+    private int size;
+
+    public ArrayList() {
+        capacity = DEFAULT_CAPACITY;
+        arrayList = (T[]) new Object[capacity];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == capacity) {
+            grow();
+        }
+        arrayList[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        if (size == capacity) {
+            grow();
+        }
+        for (int i = size; i > index; i--) {
+            arrayList[i] = arrayList[i - 1];
+        }
+        arrayList[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isIndexOutOfBounds(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        return arrayList[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (isIndexOutOfBounds(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        arrayList[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (isIndexOutOfBounds(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        final T removedElement = arrayList[index];
+        for (int i = index; i < size - 1; i++) {
+            arrayList[i] = arrayList[i + 1];
+        }
```

Comment:
Use [System.arraycopy()](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#arraycopy-java.lang.Object-int-java.lang.Object-int-int-) to move your array elements.